### PR TITLE
Testing: framework for constructing arbitrary enums for unit testing

### PR DESCRIPTION
**Pull request by wangandi**
Part of #5684

Design doc here: https://github.com/MaterializeInc/materialize/blob/d4cf3d360c88a9b1e1471e063ebd2749f030176f/doc/developer/design/20210601_build_mirrelationexpr.md

Commits in this PR have specifically organized to make it easier to read.

# What this PR does

First commit: A general framework for creating arbitrary `IR`s, which are in crates `src/lowertest` and `src/lowertest-derive`. Suggestions for alternative crate names welcome.
Second commit: Transform testing supports constructing every variant of `MirRelationExpr`, every variant of `MirScalarExpr`, and every variant of `ScalarType`. 
Third commit: Moving a unit test to using the new syntax.

# What is missing from this PR

A bunch of types of Datums/Literals cannot be constructed yet, including Literal errors.

### CODE REVIEW

**Comment by benesch** [comment]
The first commit LGTM, if you want to merge that one immediately! Then I have two minor points of structural feedback on the other commits:

1. The testing framework doesn't seem fundamental enough to belong in ore—at least not yet. Can you move it to a new crate instead?
2. Our convention is not to nest crates (see e.g. avro-derive), so `mzdecompose-derive` should live in `src` directly.

I think the rest of the review something that the optimizer team (@asenac/@frankmcsherry) will need to take on. I don't have enough context on what the pain points with the old testing framework were, and what tradeoffs were evaluated in the design of the new testing framework. My immediate reaction is that the new code introduces a good deal of complexity (a procedural macro, enum reflection, and a trip through `serde_json` that I'm having trouble following on a first skim) in order to cut down on a bit of boilerplate in the sexp parser. Personally I think I'd have just written a bit more boilerplate, but I'm not the one working with the testing framework, so I think it's a call for you + the optimizer team to make. (That's part of the rationale for confining the framework to its own crate to start, so that it's more clear where the responsibility lies.)

**Comment by wangandi** [comment]
Thanks @benesch for the structural feedback + immediate reaction. I'd just wanted to check the approach is not objectionable from an "our rust standards" viewpoint. 

Will merge first commit. 

Others should hold off on reviewing the rest until I put up a design doc (this is also when I'll add other people to the review).

**Comment by wangandi** [comment]
Finally reopening this for review! I have updated the description.

Since this PR is kind of large, I think the way to go about it is to focus on reviewing the first commit, and only actually review the second commit if the first commit is acceptable. The second commit exists to give an example of how one would use the general framework to create an IR. (Note that I intend to, in the future, move arbitrary MIR creation out of `transform/tests/test_runner.rs` into its own crate that way you can create an MIR for the `dataflow`, `expr`, and `transform` crates.)

File: src/lowertest/src/lib.rs
Comment (asenac) [inline src/lowertest/src/lib.rs:152]: Should we not try override_syntax before any default syntax?

File: src/expr/src/relation/mod.rs
Comment (asenac) [inline src/expr/src/relation/mod.rs:19]: Random thought... Should we not generate the reflection information when compiling the test config? My worry being that unused code slowing code the rest of the configs. It might be neglectable though. Out of curiosity, have you run a compile time test before and after this patch? Anyway, I'm sure it's neglectable, but asking just in case...

**Review by asenac — APPROVED** [review]
I've learned a lot of cool things from this review, including how to create proc macros and so on. Thank you!

I look forward to adding tons of tests now that we can have complex scalar expressions :)

File: src/lowertest/src/lib.rs
Comment (wangandi) [reply src/lowertest/src/lib.rs:152]: `parse_as_enum_or_struct` also calls `override_syntax` (see line 452). I'll add a comment a couple of lines up.

For enums/structs, I wanted to change what gets passed into `override_syntax`. See line 140 in the README.md for the explanation. 

The rationale is that:
The syntax supports specifying a single-arg enum/struct specification as `(<the_arg> )` or `<the_arg>` . I found it really annoying when implementing `override_syntax` for `MirRelationExpr` and `MirScalarExpr` that the code always looks like:
```
match first_arg {
  TokenTree::Ident => {..},
  TokenTree::Literal(literal) => {..},
  TokenTree::Group(group) => {
     match group.stream().into_iter().next() {
         // a repeat of the code for handling first_arg = TokenTree::Ident or Literal
     } 
  }
}
```

If we know the type is supposed to be an enum or a struct, I just want the args to the enum/struct passed to the override and not the potential outer parentheses.

File: src/expr/src/relation/mod.rs
Comment (wangandi) [reply src/expr/src/relation/mod.rs:19]: I have not tried running a compile time test, but I will. 

I originally tried using `#[cfg(test)]` to that way the macro is only generated when compiling a test, but it turns out that if you use `#[cfg(test)]`, you only generate the macro for `expr/tests` if the macro is on an enum/struct in `expr`, and the macro becomes invisible to `transform/tests`.

That said, I just had an idea - a way we could generate reflection information only at test time is to search the .rs files for the text corresponding to the struct/enum, turn the text into a `proc_macro2::TokenStream`, and then feed that `proc_macro2::TokenStream` into the existing reflection-generation method.

Comment (benesch) [reply src/expr/src/relation/mod.rs:19]: Personal take is that it's not worth worrying about. Compile time in Rust is sooo dominated by LLVM codegen that it is extremely unlikely that a proc macro could meaningfully impact compile times. 

FWIW, `walkabout` takes the approach you describe, and parses `.rs` files from source at compile time. It does this out of necessity, because it needs to consider multiple structs/enums at one time and proc macros don't allow for that. I wouldn't recommend this path if you can help it—reinventing the logic for parsing `mod` declarations sucks, and walkabout's implementation is incomplete—but that logic exists if you need it. I think `walkabout::parse_mod` is actually _exactly_ what you'd need.

Comment (wangandi) [reply src/expr/src/relation/mod.rs:19]: Is `walkabout::parse_mod` public? If not, would you be okay with making it public?

Comment (benesch) [reply src/expr/src/relation/mod.rs:19]: Yeah, that seems fine for now. One day we may want a `proc-macro-util` crate or somesuch.

**Comment by wangandi** [comment]
Merged first commit of this PR in #7118. Will post another review for the rest once I move arbitrary MIR creation out of `transform/tests/test_runner.rs`.

**Pull request closed without merging**
